### ISSUE TITLE
Implicit `any`s with `Map` even if `noImplicitAny` is enabled

### ISSUE THREAD

**Issue by Daniel15** [microsoft/TypeScript#35259]
**TypeScript Version:**  vNightly in Playground

**Search Terms:**
inferred any, implicit any, unsound

**Code**

```ts
function getStrings(): Map<string, string> {
  const val = new Map();
  val.set(42, true);
  return val;
}
```

**Expected behavior:**
I expected this to throw an error if `noImplicitAny` is enabled.

**Actual behavior:**
No errors

[Playground link](https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191120&ssl=1&ssc=1&pln=5&pc=2#code/GYVwdgxgLglg9mABAcwKZQMpQE4zMgZwAoBKALkQFkBDABwB4Cc9kAaRJ3fAPkQG8AsAChEoxBARNEAN2oAbRAF5EYVAHcqdUgG5hYmfIB0BdEQAsAJnY4QqErpFjs6ENiSy5DgL5A)

**Related Issues:**
No

**Comment by RyanCavanaugh** [comment]
Duplicate #33279

Search term: `noImplicitAny Map`

**Comment by Daniel15** [comment]
@RyanCavanaugh So is this actually expected behaviour? This seems like as big hole in TypeScript's typing. 

[Flow gets this right](https://flow.org/try/#0GYVwdgxgLglg9mABAcwKZQMpQE4zMgZwAoBKALkQFkBDABwB4Cc9kAaRJ3fAPkQG8AUIkQQETRADdqAG0QBeRGFQB3KnVIBuIZJkA6AuiIAWAEzscIVCS3Ds6ENiRTpWgL5A).

**Comment by krisdages** [comment]
I've also been burned by this kind of thing before; usually when RXJS pipe inference gets widened to `any` without my realizing it. 

Even if noImplicitAny is working as expected, 

A stricter compile option that would force assertions before passing or assigning something with an inferred `any` Type Parameter to an explicitly narrower type would be immensely helpful.  

This example with `new Map()` is relatively simple, but in more complex inference scenarios (e.g rxjs pipe chaining), it's difficult to tell when you have lost type safety, but the error potential is huge.

**Comment by krisdages** [comment]
Another source of these unexpected "any" bugs is when a conditional type resolves to "any", often unexpectedly. 

I have looked into things like dtslint to try and catch these, but they have been sort of a square peg / round hole situation.

If there are any alternatives to a compiler option that could alert me when `any` is being used when the declaration says something else, I would love to find out about them.
